awx/ui/client/src/instance-groups/instance-groups.service.js

```diff
@@ -9,33 +9,70 @@ export default
             },
             editInstanceGroups: function(url, instance_groups) {
                 Rest.setUrl(url);
-                let currentGroups = Rest.get()
-                    .then(({data}) => {
-                        return data.results.map((i) => i.id);
-                    });
+                return Rest.get()
+                    .then(res => {
+                        const { data: { results = [] } } = res;
+                        const updatedGroupIds = (instance_groups || []).map(({ id }) => id);
+                        const currentGroupIds = results.map(({ id }) => id);
 
-                return currentGroups.then(function(current) {
+                        const groupIdsToAssociate = [];
+                        const groupIdsToDisassociate = [];
+                        // loop over the array of currently saved instance group ids - if we encounter
+                        // a difference between the current array and the updated array at a particular
+                        // position, mark it and all remaining currentIds in the array for disassociation.
```

Comment:
This compares the currently saved ids and the selected ids - as soon as we encounter a difference between the two arrays, we mark _all_ remaining currently saved ids for disassociation.

Then, we build the association array in the correct order by adding any selected id that was marked for disassociation due to being out of order _and_ any ids that are new associations.

```
# scenario:
# -----------
# page is loaded with [1,2,3,4,5,6] as the currently selected tags
# user removes tag 3 from the middle
# user adds a new tag 7 to the end
# user appends tag 3 to the end

current:  [1,2,3,4,5,6]
selected: [1,2,4,5,6,7,3] 

-> idsToDisassociate = [3,4,5,6]
-> idsToAssociate = [4,5,6,7,3]
```

When sending requests to the API, we do all of the disassociations first (and wait for their completion) before doing the associations.